throttle/contract_tx_limiter_middleware.go

```diff
@@ -0,0 +1,178 @@
+package throttle
+
+import (
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/loomnetwork/go-loom"
+	"github.com/loomnetwork/go-loom/plugin/contractpb"
+	"github.com/loomnetwork/loomchain"
+	"github.com/loomnetwork/loomchain/auth"
+	udw "github.com/loomnetwork/loomchain/builtin/plugins/user_deployer_whitelist"
+	"github.com/loomnetwork/loomchain/vm"
+	"github.com/pkg/errors"
+)
+
+var (
+	ErrTxLimitReached         = errors.New("tx limit reached, try again later")
+	ErrContractNotWhitelisted = errors.New("contract not whitelisted")
+)
+
+type ContractTxLimiterConfig struct {
+	// Enables the middleware
+	Enabled bool
+	// Number of seconds each refresh lasts
+	ContractDataRefreshInterval int64
+	TierDataRefreshInterval     int64
+}
+
+func DefaultContractTxLimiterConfig() *ContractTxLimiterConfig {
+	return &ContractTxLimiterConfig{
+		Enabled:                     false,
+		ContractDataRefreshInterval: 15 * 60,
+		TierDataRefreshInterval:     15 * 60,
+	}
+}
+
+// Clone returns a deep clone of the config.
+func (c *ContractTxLimiterConfig) Clone() *ContractTxLimiterConfig {
+	if c == nil {
+		return nil
+	}
+	clone := *c
+	return &clone
+}
+
+type contractTxLimiter struct {
+	// contract_address to limiting parametres structure
+	contractToTierMap       map[string]udw.TierID
+	contractDataLastUpdated int64
+	// track of no. of txns in previous blocks per contract
+	contractToBlockTrx  map[string]*blockTxn
+	tierMap             map[udw.TierID]udw.Tier
+	tierDataLastUpdated int64
+}
+
+type blockTxn struct {
+	txn         int64
+	blockHeight int64
+}
+
+var TxLimiter *contractTxLimiter
+
+func (txl *contractTxLimiter) isAccountLimitReached(contractAddr loom.Address, curBlockHeight int64) bool {
+	blockTx, ok := txl.contractToBlockTrx[contractAddr.String()]
+	if !ok {
+		return false
+	}
+	// if execution reaches here => tierID and tier are valid
+	tierID := txl.contractToTierMap[contractAddr.String()]
+	tier := txl.tierMap[tierID]
+	if blockTx.blockHeight <= curBlockHeight-int64(tier.BlockRange) || int64(tier.MaxTx) > blockTx.txn {
+		return false
+	} else {
+		return true
+	}
+}
+
+func (txl *contractTxLimiter) updateState(contractAddr loom.Address, curBlockHeight int64) {
+	blockTx, ok := txl.contractToBlockTrx[contractAddr.String()]
+	tierID := txl.contractToTierMap[contractAddr.String()]
+	tier := txl.tierMap[tierID]
+	if !ok || blockTx.blockHeight <= curBlockHeight-int64(tier.BlockRange) {
+		// resetting the blockHeight to lower bound of range instead of curblockheight
+		rangeStart := (((curBlockHeight - 1) / int64(tier.BlockRange)) * int64(tier.BlockRange)) + 1
```

Comment:
Isn't this equivalent to `rangeStart := curBlockHeight`?

Reply:
@enlight it is lower bound of current block-range, fo for 1-10 with a block range of 10 => it is 1, for 11-20 it is 11 and so on .. @ParthDesai

Reply:
@enlight this is more on the line to store range start, instead of storing `curBlockHeight` at time of recalculation. Advantage here, is more uniform ranges, instead of ranges depending upon incoming blocks (as we could not guarantee blocks will come sequentially)